### ISSUE TITLE
Explain how to  hook in to engine

### ISSUE THREAD

**Issue by Stretto** [abulmo/amoeba#6]
Hi, I would like to use your engine to test different chess algorithm ideas I have.

Could you explain where are the key points to hook in to and what the algorithm needs to do?

The algorithm will most likely need to be ran before the built in algorithms and use it's own search space techniques.

I would like to replace the entire engine temporarily then eventually try to use it has a heuristic for my own algorithms. But I would like to use all the machinery you have developed(UCI, etc) to make it easier for me to develop. I would like to focus on the algorithm itself rather than learning the structure of the app(at this point).

e.g., how could I replace your engine with one that simply makes valid random moves? (or even invalid moves the valid move code in your engine cannot be used independently).

Thanks!

**Comment by abulmo** [comment]
The search part of the engine is in search.d file. The evaluation function is in eval.d (and the weights in weight.d). The move generator in board.d and move.d is designed to work with my own search implementation (staged move generation using some informations from the search) but can be used independently.
To replace my search logic with your own, you can look at uci.d and replace the search.* methods with your own functions. The most important functions are:
 - search.set(board): set the chess position to search
 - search.go(...): launch the search. You can use your own parameters here. 
 - search.bestmove: send the bestmove found by the search
 - search.hint: send the expected opponent answer to the bestmove: needed only if you use pondering
 - if your search is going to take some time, you have to handle the asynchronous message send by the GUI (isready, stop, ponderhit), and use the event class in your search module.
 - you can remove all what is inside the show function in file uci.d (it is here only for debugging purpose).

For an engine sending random moves I would write a class like this
```
import board, move;
import std.random;

class Search {
    Board board;
    Move bestmove;
    Mt1997 r;

    void set(in Board b) {
        board = b;
    }

    void go() {
        Moves moves;
        moves.generate(board);
        bestmove = moves[r.uniform(0, moves.length)];
   }
}

```
and remove the useless stuff in uci.d

Have a good luck

**Comment by Stretto** [comment]
Thanks. took a bit more work to get it working to use most of the original uci functionality.

The following going should be a drop in replacement for search.d. It makes random moves for the pawns, and when it does an illegal move winboard stops the game. I added some hack to bring up a windows console for better logging.

Thanks!


```
module search;


import board, move, util, eval;
import std.random, std.stdio, std.conv;
version(Windows) import core.sys.windows.windows;


// Search termination
struct Termination {
	struct Time {
		double max;
		double extra;
	}
	struct Nodes { ulong max; }
	struct Depth { int max;	}
	Time time;
	Nodes nodes;
	Depth depth;				
}



struct Option {
	Termination termination;
	int depthInit;
	size_t multiPv;
	bool easy;
	bool isPondering;
	bool verbose;

	void show() 
	{

	}
}


class Search {
    Board board;
    Move bestmove;
	Eval eval;
	shared Event event;	
	std.stdio.File logFile;
	Option option;


	void resize(size_t size) { }

	void showSetting() { }	

	void clear(in bool cleaner = true) 	{ }
    
	void set(in Board b) { board = cast(Board)b; }

	Move hint() const @property { return bestmove; }

	Move bestMove() @property { return bestmove; }

    void go(in ref Termination t, in ref Moves _moves, in bool easy = false, in int doMultiPv = 1, bool ponder = false) 
	{
		Write("go()...");
		
		Moves moves;		
        moves.generate(board);						
        //bestmove = moves[uniform(0, cast(int)moves.length)];


		auto i = 0;
		auto ms = ["", "a2a3","a2a4","b2b3","b2b4","c2c3","c2c4","d2d3","d2d4","e2e3","e2e4","f2f3","f2f4","g2g3","g2g4","h2h3","h2h4"];
		while(ms[i] == "") i = uniform(0, cast(int)ms.length);
		bestmove = toMove(ms[i]);
		Write(to!string(ms[i]));
   }
}



// Windows Console
version(Windows)
{
	static short currentY = 0;
	static int currentA = FOREGROUND_INTENSITY;
	public static void Write(string line, int attr = -1, short xLoc = -1, short yLoc = -1, bool resetXY = false) 
	{
		if (attr == -1) { attr = currentA; }
		if (resetXY) { currentY = 0; }
		if (xLoc == -1) xLoc = 0;
		if (yLoc == -1) yLoc = currentY;
	
		import core.sys.windows.windows;
		auto h = CreateFile("CONOUT$", 0x40000000, FILE_SHARE_WRITE, null, OPEN_ALWAYS, 0, null);

		short writeHeight = 1;
		short writeWidth = cast(short)line.length;

		if (h) {
			CHAR_INFO[] buf = new CHAR_INFO[writeWidth * writeHeight];
			SMALL_RECT rect = {xLoc, yLoc, cast(short)(writeWidth + xLoc), cast(short)(writeHeight + yLoc) };

			for (int i = 0; i < writeWidth; i++) {
				buf[i].Attributes = cast(ushort)attr;
				buf[i].Char.UnicodeChar = line[i];
			}

			COORD c1 = { writeWidth, writeHeight };
			COORD c2 = { 0, 0 };

		
			WriteConsoleOutput(h, buf.ptr, c1, c2, &rect);
			currentY++;
		}
	}


	shared static this()
	{
		const int WIDTH=50;
		const int HEIGHT=20;


		auto c = AllocConsole();
		auto h = GetStdHandle(c);

		HANDLE write_handle = cast(void*)h;

		SMALL_RECT window_size={0,0,WIDTH-1,HEIGHT-1};
		COORD screen_buffer={WIDTH,HEIGHT};

		SetConsoleWindowInfo(write_handle,TRUE,&window_size);
		SetConsoleScreenBufferSize(write_handle,screen_buffer);
		SetConsoleTitle("Amoeba Console");


		Write("Initializing.",FOREGROUND_INTENSITY | FOREGROUND_GREEN, 0, 0);
		Write("",FOREGROUND_INTENSITY);
	}
}  else 
{
	public static void Write(string line, int attr = -1, short xLoc = -1, short yLoc = -1, bool resetXY = false) { }
} 
```

**Comment by Stretto** [comment]
How do I get the last move that the opponent made? Can't seem to figure it out? Can't do much without knowing that info, can't seem to find it in the uci protocol docs either. ;/

**Comment by Stretto** [comment]
Is there any way to get just the last move in SAN or a simple struct? I see that board does contain the moves(if I print out board) but then I have to reverse the string to find the coordinates and piece that moved. It also seems to not update the last move properly. i.e., when I make a move for the first time, I get back the initial board. Only after the opponent makes a move do I see that I have moved also. This isn't a big deal since I know where I moved, but seems odd.
